lib/instrumental_tools/metric_script_executor.rb

```diff
@@ -0,0 +1,96 @@
+require 'benchmark'
+
+class MetricScriptExecutor
+
+  attr_reader :directory, :previous, :last_ran_at
+
+  def initialize(directory)
+    @directory   = directory
+    @previous    = {}
+  end
+
+  def can_execute_file?(path)
+    stat = File::Stat.new(path)
+    stat.executable? && file_is_owner_only?(stat)
+  end
+
+  def can_execute_in_directory?(directory)
+    stat = File::Stat.new(directory)
+    stat.directory? && file_is_owner_only?(stat)
+  end
+
+  def file_is_owner_only?(file_stat)
+    file_stat.owned? && ((file_stat.mode & 0xFFF) ^ 0O700) == 0
+  end
+
+  def run
+    process_to_output = {}
+    if can_execute_in_directory?(directory)
+      current = Dir[File.join(directory, "*")].map do |path|
+        full_path = File.expand_path(path)
+        if can_execute_file?(path)
+          stdin_r,  stdin_w  = IO.pipe
+          stdout_r, stdout_w = IO.pipe
+          stderr_r, stderr_w = IO.pipe
+
+          previous_status, previous_time, previous_output = previous[full_path]
+
+          stdin_w.write(previous_output || "")
+          stdin_w.close
+
+
+          cmd = [full_path, (previous_time || 0).to_i, (previous_status && previous_status.to_i)].compact.join(" ")
+
+          pid = Process.spawn(cmd,
+                              :chdir => File.dirname(full_path),
+                              :in    => stdin_r,
+                              :out   => stdout_w,
+                              :err   => stderr_w)
+
+          exit_status = nil
+          exec_time   = Benchmark.realtime do
```

Comment:
do you want to give a max time on these guys?  maybe even execute them in threads so that you can still be assured once every interval even if there are a lot?